src/site_apportionment/site_apportionment.py

```diff
@@ -1,98 +1,321 @@
+
+    Returns:
+        int: The number of duplicate sites.
+    """
+    site_count_df = sites_df[groupby_cols + [postcode_col]].copy()
+    site_count_df["site_count"] = site_count_df.groupby(groupby_cols + [postcode_col])[
+        postcode_col
+    ].transform("count")
+    df_duplicate_sites = site_count_df[site_count_df["site_count"] > 1]
+    num_duplicate_sites = df_duplicate_sites.shape[0]
+
+    if num_duplicate_sites:
+        SitesApportionmentLogger.info(
+            f"There are {num_duplicate_sites} duplicate sites."
+        )
+
+
+def calc_weights_for_sites(df: pd.DataFrame, groupby_cols: List[str]) -> pd.DataFrame:
+    """
+    Calculate weights for geographic sites.
+
+    The weights are calculated using the formula:
+        weight = site_percent / site_percent_total
+
+    The weights are then used to apportion expenditure of each reference across their sites.
+
+    Parameters:
+        df (pd.DataFrame): The input DataFrame.
+
+    Returns:
+        pd.DataFrame: The DataFrame with weights calculated for each site.
+    """
+    # Create a new column with cleaned site percentage values
+    df["site_percent"] = df[percent_col].fillna(0)
+    df.loc[(df.instance == 0), "site_percent"] = 0
 
     # Calculate the total percent for each reference and period
-    dfc["site_percent_total"] = dfc.groupby([ref, period])["site_percent"].transform(
+    df["site_percent_total"] = df.groupby(groupby_cols)["site_percent"].transform(
         "sum"
     )
 
-    # Filter out the rows where total percent is zero
-    dfc = dfc.copy()[dfc["site_percent_total"] != 0]
+    # Filter out rows where the total percent is zero.
+    df = df.copy().loc[df["site_percent_total"] != 0]
 
     # Compute weights
-    dfc["site_weight"] = dfc["site_percent"] / dfc["site_percent_total"]
+    df["site_weight"] = df["site_percent"] / df["site_percent_total"]
 
     # Remove unnecessary columns as they are no longer needed
-    dfc = dfc.drop(columns=["site_percent", "site_percent_total"], axis=1)
+    df = df.drop(columns=["site_percent", "site_percent_total"], axis=1)
 
-    return dfc
+    return df
+
+
+def create_cartesian_product(
+    sites_df: pd.DataFrame, category_df: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Creates a 'Cartesian product' of product classifications and sites.
+
+    'product classifications' are defined as the unique combination of civ/def,
+        product group (alpha) and product group (numeric)
+
+    Args:
+        sites_df (pd.DataFrame): The DataFrame with sites.
+        category_df (pd.DataFrame): The DataFrame with with the unique combination
+            of civ/def, product group (alpha) and product group (numeric)
+        group_cols (List[str]): The columns to group by.
+
+    Returns:
+        pd.DataFrame: The DataFrame with the Cartesian product.
+
+    Example:
+        Suppose we have the following DataFrames:
+
+        sites_df:
+            ref     site
```

Comment:
Added instance to the example so we know it belongs to sites and can be repeated if there are many product groups.

Removed reference 2 as it does not tell a story